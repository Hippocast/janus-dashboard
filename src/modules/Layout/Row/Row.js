import React from 'react';
import PropTypes from 'prop-types';

import block from '../../../helpers/bem-cn';

import './Row.css';

const row = block('j-row');
const column = block('j-col');

const propTypes = {
    children: PropTypes.node.isRequired,
    col: PropTypes.bool,
    className: PropTypes.string,
};

const wrapChild = (child, cn, className) => (
    <div
        className={cn('item').mix(className)}
    >
        {child}
    </div>
);

const wrapChildren = (children, cn) => children.map((item, index) => (
    <div
        className={cn('item')}
        key={index}
    >
        {item}
    </div>
));

const Row = ({ children, className, col, alignCenter, fullwidth }) => {
    const cn = col ? column : row;

    if (children.length > 1) {
        return (
            <div className={alignCenter ? cn({ 'centered': true })({ fullwidth }).mix(className) : cn({ fullwidth }).mix(className)}>
                { wrapChildren(children, cn) }
            </div>
        );
    }
    return wrapChild(children, cn, className);
};

Row.propTypes = propTypes;

export default Row;
