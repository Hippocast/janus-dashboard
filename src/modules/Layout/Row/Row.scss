.j-row {
    display: flex;
    flex-direction: row;
    justify-content: space-between;

    &--centered {
        text-align: center;
    }
}

.j-col {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
}
