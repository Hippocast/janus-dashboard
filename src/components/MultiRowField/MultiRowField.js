import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Field, FieldArray } from 'redux-form';

import block from '../../helpers/bem-cn';
import Row from '../../modules/Layout/Row/Row';
import Label from '../../modules/labels/Label';
import Input from '../../modules/inputs/Input';
import Hint from '../../modules/labels/Hint/Hint';
import Control from '../../components/Control/Control';

const row = block('j-row');

const propTypes = {
    name: PropTypes.string.isRequired,
    hint: PropTypes.string,
    placeholder: PropTypes.string,
    suffix: PropTypes.string,
    title: PropTypes.string,
};

class MultiRowField extends PureComponent {
    renderMembers = ({ fields, hint, placeholder, suffix, title }) => (
        <div>
            <div className={row()}>
                <Label>{ title } {placeholder}</Label>
                <Control
                    onClick={() => fields.push()}
                    icon="add"
                />
            </div>
            {
                hint &&
                    <Hint title>{ hint }</Hint>
            }
            {
                fields.map((member, index) =>
                    <Row className="double-fields" key={index} col>
                        <div className={row()}>
                            <div className={row('item')}>
                                <Field
                                    name={suffix ? `${member}.${suffix}` : `${member}`}
                                    type="text"
                                    component={Input}
                                    placeholder={placeholder ? placeholder : ''}
                                />
                            </div>
                            <div className={row('control')()}>
                                <Control
                                    onClick={() => fields.remove(index)}
                                    icon="remove"
                                />
                            </div>
                        </div>
                    </Row>
                )
            }
        </div>
    )

    render() {
        const { hint, name, placeholder, suffix, title } = this.props;

        return (
            <Row col>
                <FieldArray
                    name={`${name}`}
                    component={this.renderMembers}
                    hint={hint}
                    placeholder={placeholder}
                    suffix={suffix}
                    title={title}
                />
            </Row>
        );
    }
};

MultiRowField.propTypes = propTypes;

export default MultiRowField;
