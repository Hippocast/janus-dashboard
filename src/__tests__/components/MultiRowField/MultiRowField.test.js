import React from 'react';
import renderer from 'react-test-renderer';
import { Field, FieldArray, reduxForm } from 'redux-form';
import { createStore } from 'redux';
import { Provider } from 'react-redux';

import Row from '../../../modules/Layout/Row/Row';
import MultiRowField from '../../../components/MultiRowField/MultiRowField';

const spy = jest.fn();
const store = createStore(() => ({}));

const Decorated = reduxForm({ form: 'testForm' })(MultiRowField);

describe('MultiRowField component', () => {
    it('renders correctly', () => {
        const tree = renderer
            .create(
                <Provider store={store}>
                    <Decorated
                        name="mock-name"
                    />
                </Provider>
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
});

