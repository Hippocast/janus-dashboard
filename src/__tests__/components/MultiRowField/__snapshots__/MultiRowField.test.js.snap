// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiRowField component renders correctly 1`] = `
<div
  className="j-col__item"
>
  <div>
    <div
      className="j-row"
    >
      <label
        className="j-label"
        htmlFor={undefined}
      >
         
      </label>
      <button
        className="j-control"
        onClick={[Function]}
        type="button"
      >
        <span
          aria-label={undefined}
          className="j-icon j-icon--type-add j-control__icon"
        />
      </button>
    </div>
  </div>
</div>
`;
